Feature holiday management

lomamech opened the pull request:
feature(HolidayManagement)  …
- Implement API for Holiday Management
- Implement Integration test for Holiday Management
- Implement component bhEmployeeSelect
- Impement client side for Holiday Management
- Implement test E2E for Holiday Management
- Prevent to create for an employee holidays periods nested

reference #2292

lomamech committed:
test(Holiday Management)
- Implement test E2E for Holiday Management
- Prevent to create for an employee holidays periods nested

lomamech committed:
Fix conflict with test/data.sql

jniles commented on .env.development:
```diff
@@ -10,7 +10,7 @@ DB_NAME='bhima_test'
 SESS_SECRET='XopEn BlowFISH'
 
-DEBUG=db
+DEBUG=app
```
:+1:

lomamech committed:
fix(Holiday Management)
- Fix variable name

jniles commented on test/end-to-end/holidays/holidays.page.js:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * send back the number of holidays in the grid
+   */
+  getHolidayCount() {
```
It doesn't look like you ever use this.  Do you need it?

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -591,6 +594,7 @@
         },
         "SELECT": {
             "EMPTY": "Empty",
+            "EMPLOYEE": "Select an Employee",
```
:+1:

jniles commented on client/src/js/components/bhEmployeeSelect.js:
```diff
@@ -0,0 +1,47 @@
+
+  $ctrl.$onInit = function onInit() {    
+    // fired when an employee has been selected
+    $ctrl.onSelectCallback = $ctrl.onSelectCallback || angular.noop;
```
Since this is a required binding, it shouldn't ever be null.  There is no need for this code.

jniles commented on client/src/js/components/bhEmployeeSelect.js:
```diff
@@ -0,0 +1,47 @@
+    Employees.read()
+      .then(function (employees) {
+        $ctrl.employees = employees;
+      });
```
This should probably include `Notify.handleError` to catch HTTP errors.

jniles commented on client/src/js/components/bhEmployeeSelect.js:
```diff
@@ -0,0 +1,47 @@
+  };
+
+  // fires the onSelectCallback bound to the component boundary
+  $ctrl.onSelect = function ($item, $model) {
```
There's no need to have `$model` defined here.

jniles commented on client/src/modules/holidays/holidays.js:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  // create a new Holiday
+  function createHoliday() {
```
Can you just use a `ui-sref` instead of defining JavaScript functions?  Both this and edit could benefit from it.

jniles commented on client/src/modules/holidays/holidays.js:
```diff
@@ -0,0 +1,105 @@
+
+HolidayManagementController.$inject = [
+  'HolidayService', 'ModalService',
+  'NotifyService', 'uiGridConstants', '$state', 'SessionService',
```
Do you use `SessionService`?  It doesn't look like it from here.  Could you remove it?

jniles commented on client/src/modules/holidays/holidays.routes.js:
```diff
@@ -0,0 +1,42 @@
+    .state('holidays.create', {
+      url : '/create',
+      params : {
+        holiday : { value : null },
```
Do you ever use `$state.params.holiday`?   I couldn't find it.  If not, go ahead any remove these from both the edit and create state.

jniles commented on client/src/modules/holidays/holidays.service.js:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ * @class HolidayService
+ * @extends PrototypeApiService
```
:+1:

jniles commented on client/src/modules/holidays/modals/holiday.modal.html:
```diff
@@ -0,0 +1,60 @@
+      <bh-date-interval
+        label="FORM.LABELS.PERIODES"
+        date-id="holiday-date"
+        validation-trigger="HolidayModalCtrl.$submitted"
```
I think this should be `HolidayForm.$submitted`.  The controller never has a `$submitted` property set on it.

jniles commented on client/src/modules/templates/bhEmployeeSelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+<div ng-form="{{ $ctrl.name }}" bh-employee-select ng-model-options="{ updateOn: 'default' }">
+  <div
+    class="form-group"
+    ng-class="{ 'has-error' : EmployeeForm.$$parent.$submitted && EmployeeForm.employee_id.$invalid }">
```
`$$parent` doesn't exist on the form - it should be `$$parentForm`.  [See this answer for more info](https://stackoverflow.com/a/31984524/6251968)

jniles commented on client/src/modules/templates/bhEmployeeSelect.tmpl.html:
```diff
@@ -0,0 +1,24 @@
+      </ui-select-choices>
+    </ui-select>
+
+    <div class="help-block" ng-messages="EmployeeForm.employee_id.$error" ng-show="EmployeeForm.$$parent.$submitted">
```
Same here for `$$parentForm`.

jniles commented on server/controllers/admin/holidays.js:
```diff
@@ -0,0 +1,147 @@
+  checkHoliday(data)
+    .then((record) => {
+      if (record.length) {
+        throw new Unauthorized('Holiday Nested.', 'ERRORS.HOLIDAY_NESTED');
```
I don't think this should be unauthorized (403).  Probably a `BadRequest` error is better (400).

jniles commented on server/controllers/admin/holidays.js:
```diff
@@ -0,0 +1,147 @@
+  checkHoliday(req.body)
+    .then((record) => {
+      if (record.length > 1) {
+        throw new Unauthorized('Holiday Nested.', 'ERRORS.HOLIDAY_NESTED');
```
Same here - this should be `BadRequest`.  The user inserted bad information

jniles commented on server/controllers/admin/holidays.js:
```diff
@@ -0,0 +1,147 @@
+    .then((row) => {
+    // if nothing happened, let the client know via a 404 error
+      if (row.affectedRows === 0) {
+        throw new NotFound(`Could not find a Holiday with id ${req.params.id}`);
```
:+1:

jniles commented on server/controllers/admin/holidays.js:
```diff
@@ -0,0 +1,147 @@
+        throw new NotFound(`Could not find a Holiday with id ${req.params.id}`);
+      }
+
+      res.status(204).json();
```
It would be better to use `res.sendStatus(204)` since this response isn't actually JSON.

jniles commented on server/models/bhima.sql:
```diff
@@ -29,7 +29,7 @@ INSERT INTO unit VALUES
   (29,  'Patient Group','TREE.PATIENT_GRP','',1,'/modules/patients/groups/','/patients/groups'),
   (48,  'Service Management','TREE.SERVICE','',1,'modules/services/','/services'),
-  (57,  'Payroll','TREE.PAYROLL','',0,'modules/payroll/','/payroll/'),
+  (57,  'Human Ressources','TREE.HUMAN_RESSOURCES','',0,'modules/human_ressources/','/human_ressources/'),
```
`resources` should only have one `s`.

jniles commented on test/integration/holidays.js:
```diff
@@ -0,0 +1,84 @@
+ *
+ * This test suite implements full CRUD on the /holidays  HTTP API endpoint.
+ */
+describe('(/holidays) The /holidays  API endpoint', function () {
```
Could you make sure this integration test suite respects ESLint?

jniles reviewed (changes requested):
@lomamech  great work porting this!  I couldn't have been easy.  I've made a few clean up suggestions.

lomamech committed:
refactor(Holiday)
- Remove not necessary code
- Add catch HTTP errors in Employee Component
- Remove not necessary function
- Use ui-sref for location
- Sanitaze holidays.routes
- Fix, Update and Improve bhEmployeeSelect
- Use badRequest insteqd Unauthorized
- Fix test integration of holiday with eslint

lomamech committed:
Fix Conflict

lomamech committed:
Resolve conflict

lomamech committed:
Fix conflict
